packages/git/src/browser/git-contribution.ts

```diff
@@ -772,129 +555,161 @@ export class GitContribution implements
             const uri = ref.editor.uri;
-            const change = this.gitCommands.findChange(uri);
-            if (change && this.gitCommands.getUriToOpen(change).toString() !== uri.toString()) {
+            const change = provider.findChange(uri);
+            if (change && provider.getUriToOpen(change).toString() !== uri.toString()) {
                 const selection = ref.editor.selection;
                 return { change, options: { selection, widgetOptions: { ref } } };
             }
         }
         return undefined;
     }
 
-    protected hasMultipleRepositories(): boolean {
-        return this.repositoryTracker.allRepositories.length > 1;
-    }
+    protected async updateStatus(): Promise<void> {
+        const scmProvider = this.repositoryProvider.selectedScmProvider;
+        if (!scmProvider) {
+            return;
+        }
+        const status = this.repositoryTracker.selectedRepositoryStatus;
+        if (status) {
+            await scmProvider.setStatus(status);
+        }
 
-    protected updateSyncStatusBarEntry(repositoryUri: string | undefined): void {
-        const entry = this.getStatusBarEntry();
-        if (entry && repositoryUri) {
-            const scmProvider = this.scmProviders.find(provider => provider.rootUri === repositoryUri);
-            if (scmProvider) {
-                (scmProvider as ScmProviderImpl).fireChangeStatusBarCommands([{
-                    id: GitContribution.GIT_SYNC_STATUS,
-                    text: entry.text,
-                    tooltip: entry.tooltip,
-                    command: entry.command,
-                }]);
+        const statusBarCommands: ScmCommand[] = [];
+
+        const checkoutCommand = this.getCheckoutStatusBarCommand();
+        if (checkoutCommand) {
+            statusBarCommands.push(checkoutCommand);
+        }
+
+        const syncCommand = this.getSyncStatusBarCommand();
+        if (syncCommand) {
+            statusBarCommands.push(syncCommand);
+        }
+        scmProvider.statusBarCommands = statusBarCommands;
+    }
+    protected getCheckoutStatusBarCommand(): ScmCommand | undefined {
+        const scmProvider = this.repositoryProvider.selectedScmProvider;
+        if (!scmProvider) {
+            return undefined;
+        }
+        const status = scmProvider.getStatus();
+        if (!status) {
+            return undefined;
+        }
+        const branch = status.branch ? status.branch : status.currentHead ? status.currentHead.substring(0, 8) : 'NO-HEAD';
+        let dirty = '';
+        if (status.changes.length > 0) {
+            const conflicts = this.hasConflicts(status.changes);
+            const staged = this.allStaged(status.changes);
+            if (conflicts || staged) {
+                if (conflicts) {
+                    dirty = '!';
+                } else if (staged) {
+                    dirty = '+';
+                }
+            } else {
+                dirty = '*';
             }
```

Comment:
for completeness: this is missing the case `*+` to indicate both kinds.

Reply:
well then even `*+!` should be possible, will update

Reply:
should be good now